### ISSUE TITLE
[Feature request]: Move cache directory out of individual app directories

### ISSUE THREAD

**Issue by hyuri** [flatpak/flatpak#5407]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

I would like to suggest moving the cache folder out of individual app directories. Perhaps to `~/.var/cache`, alongside `~/.var/app`.

This makes it much easier to backup app preferences/settings, skipping the often gigabytes of cache files that need not be backed up — as you know.

So, going from:

`~/.var/app/`
- `org.name.app1/cache`
- `org.name.app2/cache`
- `org.name.app3/cache`

to:

`~/.var/app/`
- `org.name.app1`
- `org.name.app2`
- `org.name.app3`

`~/.var/cache/`
- `org.name.app1`
- `org.name.app2`
- `org.name.app3`

**Comment by simpz** [comment]
Arghh, sadly imagined this might be easy to do (config wise).

This is pretty essential in a corporate setting, things like browser caches should be able to be moved to fast local storage (often SSD) and not be clogging NFS homedir servers with this stuff (both in storage space and I/O usage)

**Comment by simpz** [comment]
I tried with firefox with what I thought would work from documentation, to move just the cache for my user:
flatpak override --user --filesystem=xdg-cache/mnt/local/.var/org.mozilla.firefox/cache:create org.mozilla.firefox

This didn't work, still went to ~/.var on the NFS homedir

I even tried moving the whole thing to local storage (/mnt/local is local storage)
flatpak override --user --filesystem=/mnt/local/simpz/.var/org.mozilla.firefox/cache:create org.mozilla.firefox

It still stuck to  ~/.var/app/org.mozilla.firefox on the NFS mount.

The only thing I could get to work was revert everything:
flatpak override --user --reset org.mozilla.firefox

ln -s /mnt/local/simpz/.var/org.mozilla.firefox ~/.var/app/org.mozilla.firefox 

Am I missing something?

**Comment by smcv** [comment]
> flatpak override --user --filesystem=xdg-cache/mnt/local/.var/org.mozilla.firefox/cache:create org.mozilla.firefox

This does what it says: it creates `~/.cache/mnt/local/.var/org.mozilla.firefox/cache` (on the host, and shares it with the sandbox). But there's no reason why Firefox would choose to write to that directory in preference to `~/.var/org.mozilla.firefox/cache`. Neither Flatpak nor Firefox can read your mind and "know what you mean" just because the paths look sort of vaguely similar.

>  flatpak override --user --filesystem=/mnt/local/simpz/.var/org.mozilla.firefox/cache:create org.mozilla.firefox

Similarly this creates `/mnt/local/simpz/.var/org.mozilla.firefox/cache` and shares it with the sandbox, but there's no reason why Firefox would write there.

**Comment by simpz** [comment]
Okay but I assumed that the container would use the setting for xdg-cache and this had some special significance i.e. used for cache locations by the application (as good practice when they are built). If not what is the point of the,xdg-cache directive to filesystem:
 --filesystem=xdg-cache/mnt/local/.var/org.mozilla.firefox/cache:create org.mozilla.firefox

Sorry I'm new to flatpak.

**Comment by smcv** [comment]
The purpose of `xdg-cache` is so that app authors can say "whatever the user's `$XDG_CACHE_HOME` or `~/.cache` is" in their manifest, without needing to know what specific path you might have set that to.

For instance an IDE that contains compilers might have `--filesystem=xdg-cache/ccache`, so that it will share `~/.cache/ccache` (or `$XDG_CACHE_HOME/ccache`, if set) with your host system.

**Comment by smcv** [comment]
> sadly imagined this might be easy to do (config wise)

It will not be easy to do unless someone (perhaps you) contributes code to Flatpak that adds an option similar to the one suggested here.

**Comment by simpz** [comment]
Maybe the wrong place to continue the discussion, but is it not simply a matter of getting app packagers to follow the convention of xdg-cache, xdg-config etc and obey the request to place these classes of data in these locations?

**Comment by smcv** [comment]
No, the applications that run inside the sandbox don't have control over how Flatpak arranges directories from outside the sandbox to set it up.

**Comment by lectrode** [comment]
Per-user login script to redirect the individual app cache folders to subfolders under host `$XDG_CACHE_HOME`. Useful for easy exclusion from backup software, and (if you have your XDG_CACHE_HOME mapped to tmpfs) increased performance, and decreased wear on solid state drives.

Note: Ignores runtimes; only redirects caches for "apps" (`flatpak list --columns=application --app`).
Note: Script is intended to run in bash. Written/tested on Arch Linux

```
#!/bin/bash

# Replaces "cache" folders for flatpak apps with a symlink to "${XDG_CACHE_HOME}/$cachefol/<appid>"
# Sets app-specific access overrides at the user level to grant permission to the new caches

############
# settings #
############

cachefol="xs-fpcache"
thisapp="mvfpcache"
errfile="xs-${thisapp}-err.txt"

#############
# functions #
#############

msg_yel(){ (echo -e "$@";echo;echo)>>"$errfile"; notify-send -i dialog-warning "[$thisapp]" "$@" ; }

symfol(){
#$1=original folder path (to be replaced with symlink)
#$2=new destination folder path
  [[ -d "$(dirname "$1")" ]] || return
  if [[ ! -d "$2" ]]; then mkdir -p "$2"; sync; fi
  if [[ ! "$1" = "" ]] && { [[ -e "$1" ]] || [[ -L "$1" ]]; }; then
    #check if symlink already exists and points to destination, otherwise remove
    if [[ "$(realpath "$1")" = "$(realpath "$2")" ]]; then return
    else rm -rf "$1"; sync; fi
  fi
  ln -s "$2" "$1" || msg_yel "failed to create symlink:\n$1\n$2"
}

##########
#  main  #
##########

pushd "$HOME" || exit
if [[ "$XDG_CACHE_HOME" = "" ]] || [[ ! -d "$XDG_CACHE_HOME" ]]; then
  XDG_CACHE_HOME="${HOME}/.cache"
fi

errfile="${XDG_CACHE_HOME}/$errfile"
[[ -f "$errfile" ]] && rm "$errfile"

while read -r f; do
  if ! flatpak override --show "$f" --user|grep -E '^filesystems='|grep -F "xdg-cache/$cachefol/$f" >/dev/null; then
    if ! flatpak override --filesystem="xdg-cache/$cachefol/$f" "$f" --user; then
      msg_yel "failed to set cache permissions:\n$f"
      continue
    fi
  fi
  symfol "${HOME}/.var/app/$f/cache" "${XDG_CACHE_HOME}/$cachefol/$f"
done <<< "$({ flatpak list --columns=application --app;flatpak list --columns=application --app --user; }|grep -F '.')"


exit 0

```

**Comment by lectrode** [comment]
with flatpak 1.18.1, the symlink workaround now fails with:

    error: absolute symlink not allowed for RESOLVE_BENEATH: Invalid cross-device link

this is the case even with the app having explicit permissions allowed to access the symlink target directory
